lib/src/gotrue_client.dart

```diff
@@ -687,7 +687,15 @@ class GoTrueClient {
       );
       return completer.future;
     } catch (error, stack) {
-      completer.completeError(error, stack);
+      if (error is AuthException) {
+        if (error.message == 'Invalid Refresh Token: Refresh Token Not Found') {
+          _notifyAllSubscribers(AuthChangeEvent.tokenRefreshFailed);
+          await signOut();
+          completer.complete(AuthResponse(session: null, user: null));
+        }
+      }
+      if (!completer.isCompleted) completer.completeError(error, stack);
+
```

Comment:
What do you think about instead of introducing a new `AuthChangeEvent` for failed token refresh, we add error to `_onAuthStateChangeController` so that supabase_flutter can listen to the error using `Supabase.instance.auth.onAuthStateChanged.listen().onError()`? That way the users can handle this error same way too.

```suggestion
      _onAuthStateChangeController.addError(error, stack);
      if (error is AuthException) {
        if (error.message == 'Invalid Refresh Token: Refresh Token Not Found') {
          await signOut();
          completer.complete(AuthResponse(session: null, user: null));
        }
      }
      if (!completer.isCompleted) completer.completeError(error, stack);

```

I created an PR on supabase_flutter here so that it can pick up the errors on gotrue-dart and print it.
https://github.com/supabase/supabase-flutter/pull/416

Also going beyond the scope of this PR, I think it makes sense to add all possible auth errors to `_onAuthStateChangeController` so that any auth errors can be handled just by listening to `onAuthStateChanged`, but what do you think?

Reply:
I like the idea, but I see a contradiction in returning an empty `AuthResponse` to the completer, but emitting an error to the stream. I don't think that's a good idea.

>  I think it makes sense to add all possible auth errors to _onAuthStateChangeController so that any auth errors can be handled just by listening to onAuthStateChanged

Which exact errors do you mean? Even errors from a signin or signup method?

Reply:
>I like the idea, but I see a contradiction in returning an empty AuthResponse to the completer, but emitting an error to the stream. I don't think that's a good idea.

Great point! My thought process for this implementation was that even if we throw errors on the `completer`, the user will not be able to catch it and handle it, and it just gets annoying. There still should be some way to handle errors, so we expose the error through the `onAuthStateChanged` listener. What do you think?

> Which exact errors do you mean? Even errors from a signin or signup method?

No, not the errors from signIn or signup methods, but just mostly those that supabase_flutter users have no way to catch, like errors emitted from `recoverSession`. I guess for this one, there's no reason to wait for it, huh? Let me just create a PR right now.

Reply:
From a point of view of only the gotrue client, I think `recoverSession` should still throw. It's weird if you get the error only via an extra stream. So maybe the catch to silent the error should happen in supabase-flutter. `initialSession` is definitely a problem here, but removing it would be a breaking change. I've created https://github.com/supabase/supabase-flutter/pull/418 now